ARKitExample/ViewController.swift

```diff
@@ -0,0 +1,219 @@
+        super.viewWillDisappear(animated)
+    }
+    
+    override func didReceiveMemoryWarning() {
+        super.didReceiveMemoryWarning()
+        // Release any cached data, images, etc that aren't in use.
+    }
+
+    @objc func displayLinkDidFire() {
+        // Our capturer polls the ARSCNView's snapshot for processed AR video content, and then copies the result into a CVPixelBuffer.
+        // This process is not ideal, but it is the most straightforward way to capture the output of SceneKit.
+        let myImage = self.sceneView.snapshot
+        let imageRef = myImage().cgImage
+
+        if (imageRef == nil) {
+            return
+        }
+
+        // As a TVIVideoCapturer, we must deliver CVPixelBuffers and not CGImages to the consumer.
+        let pixelBuffer = self.copyPixelbufferFromCGImageProvider(image: imageRef!)
+
+        self.frame = TVIVideoFrame(timestamp: Int64((displayLink?.timestamp)! * 1000000),
+                                   buffer: pixelBuffer,
+                                   orientation: TVIVideoOrientation.up)
+        self.consumer?.consumeCapturedFrame(self.frame!)
+    }
+
+    /**
+     * Copying the pixel buffer took ~0.026 - 0.048 msec (iPhone 7 Plus).
+     * This pretty fast but still wasteful, it would be nicer to wrap the CGImage and use its CGDataProvider directly.
+     **/
+    func copyPixelbufferFromCGImageProvider(image: CGImage) -> CVPixelBuffer {
+        let dataProvider: CGDataProvider? = image.dataProvider
+        let data: CFData? = dataProvider?.data
+        let baseAddress = CFDataGetBytePtr(data!)
+
+        /**
+         * We own the copied CFData which will back the CVPixelBuffer, thus the data's lifetime is bound to the buffer.
+         * We will use a CVPixelBufferReleaseBytesCallback callback in order to release the CFData when the buffer dies.
+         **/
+        let unmanagedData = Unmanaged<CFData>.passRetained(data!)
+        var pixelBuffer: CVPixelBuffer? = nil
+        let status = CVPixelBufferCreateWithBytes(nil,
+                                                  image.width,
+                                                  image.height,
+                                                  TVIPixelFormat.format32BGRA.rawValue,
+                                                  UnsafeMutableRawPointer( mutating: baseAddress!),
+                                                  image.bytesPerRow,
+                                                  { releaseContext, baseAddress in
+                                                    let contextData = Unmanaged<CFData>.fromOpaque(releaseContext!)
+                                                    contextData.release() },
+                                                  unmanagedData.toOpaque(),
+                                                  nil,
+                                                  &pixelBuffer)
+
+        if (status != kCVReturnSuccess) {
+            return nil as CVPixelBuffer!;
```

Comment:
This seems problematic? We are saying that this method returns a `CVPixelBuffer`, but in this pathway we are returning `nil`? I am surprised that the compiler doesn't get angry about this. Seems like make we should just make the return value an optional and not do the force unwrapping.

Reply:
Good point, I think making the return optional is the right way to go. Silly copy/paste of similar code from ScreenCapturerExample which didn't have the same nullability requirements.